lib/Interpreter/CppInterOp.cpp

```diff
@@ -1027,61 +1056,78 @@ namespace Cpp {
   }
 
   TCppFunction_t
-  BestTemplateFunctionMatch(const std::vector<TCppFunction_t>& candidates,
+  BestOverloadFunctionMatch(const std::vector<TCppFunction_t>& candidates,
                             const std::vector<TemplateArgInfo>& explicit_types,
                             const std::vector<TemplateArgInfo>& arg_types) {
-
-    for (const auto& candidate : candidates) {
-      auto* TFD = (FunctionTemplateDecl*)candidate;
-      clang::TemplateParameterList* tpl = TFD->getTemplateParameters();
-
-      // template parameter size does not match
-      if (tpl->size() < explicit_types.size())
-        continue;
-
-      // right now uninstantiated functions give template typenames instead of
-      // actual types. We make this match solely based on count
-
-      const FunctionDecl* func = TFD->getTemplatedDecl();
+    auto& S = getSema();
+    auto& C = S.getASTContext();
 
 #ifdef CPPINTEROP_USE_CLING
-      if (func->getNumParams() > arg_types.size())
-        continue;
-#else // CLANG_REPL
-      if (func->getMinRequiredArguments() > arg_types.size())
-        continue;
+    cling::Interpreter::PushTransactionRAII RAII(&getInterp());
 #endif
 
-      // TODO(aaronj0) : first score based on the type similarity before forcing
-      // instantiation.
-
-      TCppFunction_t instantiated =
-          InstantiateTemplate(candidate, arg_types.data(), arg_types.size());
-      if (instantiated)
-        return instantiated;
-
-      // Force the instantiation with template params in case of no args
-      // maybe steer instantiation better with arg set returned from
-      // TemplateProxy?
-      instantiated = InstantiateTemplate(candidate, explicit_types.data(),
-                                          explicit_types.size());
-      if (instantiated)
-        return instantiated;
-
-      // join explicit and arg_types
-      std::vector<TemplateArgInfo> total_arg_set;
-      total_arg_set.reserve(explicit_types.size() + arg_types.size());
-      total_arg_set.insert(total_arg_set.end(), explicit_types.begin(),
-                           explicit_types.end());
-      total_arg_set.insert(total_arg_set.end(), arg_types.begin(),
-                           arg_types.end());
-
-      instantiated = InstantiateTemplate(candidate, total_arg_set.data(),
-                                         total_arg_set.size());
-      if (instantiated)
-        return instantiated;
+    llvm::SmallVector<Expr*> Args;
+    for (auto i : arg_types) {
+      QualType Type = QualType::getFromOpaquePtr(i.m_Type);
+      ExprValueKind ExprKind = ExprValueKind::VK_PRValue;
+      if (Type->isReferenceType())
+        ExprKind = ExprValueKind::VK_LValue;
+      Args.push_back(new (getSema().getASTContext())
```

Comment:
Construct the Expr on the stack and pass it as a reference. This way if we start using something outside of the types and expression values valgrind will be able to tell us (upon future changes of the implementation in clang). To do this you will need an `std::vector<Expr> Exprs` with the same lifetime as `Args` and `Args` taking `Exprs` addresses elementwise.

Reply:
I seem to remember that we could create an expression on the stack but apparently I am confusing myself. We should probably put it in some scratch area, call placement new there and then destroy it...

Reply:
Not sure how I would go about destroying it. The default `delete` operator is marked unreachable (reference: https://github.com/llvm/llvm-project/blob/release/19.x/clang/include/clang/AST/Stmt.h#L106-L109). And the `delete` operators defined are empty (reference: https://github.com/llvm/llvm-project/blob/release/19.x/clang/include/clang/AST/Stmt.h#L1289-L1293).
I can call into one of these `delete` operators if required?